### PR TITLE
Cat state tyler

### PR DESCRIPTION
**Pull request by tylergcowan**
Qiskit cat state DQFT implementation (must be tested fully)

### CODE REVIEW

File: qiskit/cluster.py
Comment (brunorijsman) [inline qiskit/cluster.py:10]: Nitpick: "flag" is a very generic name. I think we should call it "method" or something more clear. I will not make this (or any other comments) a blocker - you can fix (or not fix, as you choose, we have bigger fish to fry before end of day today) later.

Comment (brunorijsman) [reply qiskit/cluster.py:10]: As a rule, I never comment on code formatting. Instead, I create a CI/CD pipeline that runs "black" (https://github.com/psf/black) as part of the commit pipeline, and I never even look at a pull request unless black approves it.

But since I have not yet added a CI/CD pipeline, I will just point out there are some code formatting inconsistencies (e.g. no space between , and flag, and no spaces around = below). Once again, I will not let this block the merge. We can fix (or not) later.

File: qiskit/cluster.py
Comment (brunorijsman) [inline qiskit/cluster.py:33]: We could the common part (the main register, the entanglement register, and the classical register) outside the if statement instead of copy-pasting it in both methods.

File: qiskit/cluster.py
Comment (brunorijsman) [inline qiskit/cluster.py:21]: An enum (https://docs.python.org/3/library/enum.html) might be cleaner than a string. If you use a string, you may want to throw an exception if it has an unexpected value.

File: qiskit/cluster.py
Comment (brunorijsman) [inline qiskit/cluster.py:61]: Please remove debug prints.

File: qiskit/cluster.py
Comment (brunorijsman) [inline qiskit/cluster.py:59]: What does _tp stand for? Perhaps a more clear name?

File: qiskit/cluster.py
Comment (brunorijsman) [inline qiskit/cluster.py:151]: I understand what you are doing here, but it may be clearer if you store the method in an attribute and make this conditional on method being set to teleport...

File: qiskit/cluster.py
Comment (brunorijsman) [inline qiskit/cluster.py:215]: This check should probably be done earlier when the constructor is called.

File: qiskit/cluster.py
Comment (brunorijsman) [inline qiskit/cluster.py:197]: If you store the flag (method) as in a class attribute, you can avoid passing it around in all these function calls.

File: qiskit/cluster.py
Comment (tylergcowan) [reply qiskit/cluster.py:10]: I will change flag to method

File: qiskit/cluster.py
Comment (tylergcowan) [reply qiskit/cluster.py:33]: Agreed

File: qiskit/cluster.py
Comment (tylergcowan) [reply qiskit/cluster.py:61]: will do

**Review by brunorijsman — APPROVED** [review]

File: qiskit/cluster.py
Comment (tylergcowan) [reply qiskit/cluster.py:59]: tp stands for teleport. I wanted to avoid making the function name too long, but if that isn't clear it should be teleport.

File: qiskit/cluster.py
Comment (tylergcowan) [reply qiskit/cluster.py:151]: This was a quick fix, I'll circle back to this

File: qiskit/cluster.py
Comment (tylergcowan) [reply qiskit/cluster.py:197]: definitely, I'll make this change

**Pull request merged**
